design-documents/graph-ql/nullability.md

```diff
@@ -0,0 +1,159 @@
+Making fields non-nullable alleviates some of these issues for clients.
+
+## Why nullable?
+
+When a field's resolver has an error, GraphQL requires that the value of the field be set to `null`. But, if a field is non-nullable, a `null` value would break that contract.
+
+Instead, an error in a resolver for a non-nullable field propagates up to the nearest nullable ancestor field. This can end up causing a client to lose more data farther up the response tree, even though that data may have been resolved without errors.
+
+### Example: Bad use of non-nullable field
+
+In this example, a client wants to query for data to render a product details page. But, because the `related_products` field is non-nullable, the client will lose access to data like `name` and `price` if `related_products` throws any errors.
+
+Ideally, a UI would get the critical product data, render the page, and just not render the "Related Products" display.
+
+**Schema**:
+```graphql
+type Query {
+   product(id: ID): Product
+}
+
+type Product {
+	id: ID!
+	name: String!
+    price: Money!
+    description: String!
+    url: String!
+    related_products: [RelatedProduct!]!
+}
+```
+**Query**:
+```graphql
+query ProductDetailsPage($id: ID) {
+	product(id: $id) {
+       id
+       name
+       price
+       description
+       # related_products can't be null. When the backing service is
+       # down/unreachable at the time this query runs, we can't fulfill
+       # the contract
+       related_products {
+          id
+          name
+          price
+          url
+       }
+    }
+}
+```
+
+**Response**
+
+Note that no data was obtained by the client for `product`
+```js
+{
+   "data": {
+      "product": null
+   },
+   "errors": [
+      // details about failure in related_products field populated here
+	]
+}
+```
+
+## Backwards Compatibility
+
+- It is a safe, non-breaking change to move a field from nullable to non-nullable
+- It is an unsafe, breaking change to move a field from non-nullable to nullable
+
+With this in mind, when in doubt, the safer route is to make a field nullable, with the agreement that we'll iterate and improve schemas as we get feedback on usage from clients.
+
+## Recommendations
+
+### Scalars
+
+If the field's value is obtained from a backing data source, and that value is required in the backing data source, then the field in the GraphQL schema should be non-nullable
+
+#### Example
+```graphql
+type Product {
+   id: ID! # Products service would not allow a product without an ID, so id is non-nullable
+   color: String # Backing products DB does not require color, so it's nullable
+}
+```
+
+### Object Types
+
+Object type fields are frequently used to create joins between types. Object type fields should be nullable when the backing data is likely to live in a different service from the parent object type. This rule will require more thought since it depends where we think domain boundaries will be drawn as we decompose to services.
```

Comment:
From the problem statement, it looks like the rule should be "field may be nullable if it is not critical for the parent object". So, if it's ok to have partial result returned in case child object is not available, then this child object should be nullable. If the parent object will lose integrity in case child object is unavailable, then the child should be non-nullable.
Storage may change with changes in design, but scenarios are usually the same doesn't matter what's the underlying design of the system.

Reply:
This is a great point. I re-wrote this document a few times last night, and didn't notice towards the end of the document when I wrote this line that it would apply to all fields, not just list fields:

> A list _field_ should be non-nullable when its absence would make the parent type fairly useless.

I'll refactor the document a bit to clean things up. Thanks for the useful feedback!

Reply:
@buskamuza I pushed an update to reword things - can you give it another read?